### ISSUE TITLE
Entity Splitting: allow optional split table

### ISSUE THREAD

**Issue by AndriySvyryd** [dotnet/efcore#27974]
Split off from #620

This would skip creating a row in the table mapped to the split fragment if all mapped properties are `null`

Add a warning when this is changed in a migration

**Comment by roji** [comment]
This issue is in the Backlog milestone. This means that it is not going to happen for the 8.0 release. We will re-assess the backlog following the 8.0 release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by anasik** [comment]
I put together an initial implementation of this. I am actually ready to make a PR but the guidelines say to get it approved first. 

Approach:
- `SplitTableBuilder`/`OwnedNavigationSplitTableBuilder` gain `IsOptional(bool optional = true)`. A fragment marked
  optional only gets a row once at least one of its non-key properties is non-null.
- Model validation requires every non-key property mapped exclusively to an optional fragment to be nullable
  (store-object-aware, via `IsColumnNullable`) — the fragment's key/FK column stays non-nullable, so migrations
  keep generating a normal identifying FK, nothing becomes nullable at the schema level.
- Queries translate an optional fragment's join as `LEFT JOIN` instead of `INNER JOIN` (required fragments are
  unaffected, and mixed required/optional fragments on the same entity work).
- SaveChanges on SQL Server does the insert/update/delete atomically in one round trip — an `UPDATE`, then a
  conditional `INSERT` only if it affected zero rows and the payload isn't all-null (no `MERGE`, no separate
  `SELECT` first) — and preserves normal `DbUpdateConcurrencyException` semantics when a concurrency token is
  configured on the fragment.
- Other relational providers get a clear `NotSupportedException` if you try to modify/delete an entity with an
  optional fragment, rather than incorrect SQL. Inserting a new entity with an all-null optional payload works
  on every provider, since that decision doesn't require provider-specific SQL.

**Not done yet:** the migration-time warning this issue also asks for (flagging when a fragment's
`IsOptional` changes between migrations). Wanted to get feedback on the runtime behavior/API shape first rather
than build that on top of an unreviewed approach — happy to add it if the direction here looks right.

Given the size of this, wanted to check in before assuming the approach (naming, SQL-Server-only scope for
writes, etc.) is what you'd want — let me know if anything should go differently.


@roji @AndriySvyryd

**Comment by AndriySvyryd** [comment]
Looks good overall.

On the Metadata side you can store this configuration in fk.IsRequiredDependent on the linking FK.

> SaveChanges on SQL Server does the insert/update/delete atomically in one round trip — an UPDATE, then a
conditional INSERT only if it affected zero rows and the payload isn't all-null (no MERGE, no separate
SELECT first)

This shouldn't be necessary. You can examine the original values for all the properties to determine whether to issue an UPDATE or INSERT. And this wouldn't be provider-specific, so everything can be implemented at the Relational level.

Just note that the earliest this could be shipped is in 12.0.0.

**Comment by anasik** [comment]
On `IsRequiredDependent`: 

The linking FK `EntitySplittingConvention` creates only exists after `ProcessModelFinalizing` runs, which is after `OnModelCreating` (and therefore after `IsOptional()`) has already completed so at the point `IsOptional()` is called there's no FK yet to set `IsRequiredDependent` on. 

Would the intent be for `IsOptional()` to stash the value temporarily (similar to what I have now) and have `EntitySplittingConvention` promote it onto the FK at finalization time? 

Also, currently EntitySplittingConvention creates only one linking FK for the whole entity type. Should I make it create one per fragment so mixed required/optional fragments on the same entity can each carry their own value?

Furthermore, `RelationalModel.PopulateRowInternalForeignKeys` explicitly excludes self-referencing FKs so `tableMapping.Table.GetRowInternalForeignKeys` will not even see it. 

All of this put together makes me see little benefit to moving optionality onto the FK. At this point wouldn't it be simpler to keep IsOptional() as its own flag?

**Comment by anasik** [comment]
> Just note that the earliest this could be shipped is in 12.0.0.

One more question: It makes sense that we can't ship this before 12.0.0 but then what branch should I make the PR to?

@AndriySvyryd

**Comment by AndriySvyryd** [comment]
> All of this put together makes me see little benefit to moving optionality onto the FK. At this point wouldn't it be simpler to keep IsOptional() as its own flag?

You are right, there's no immediate benefit, filed https://github.com/dotnet/efcore/issues/38746 to do it as a separate change.

> One more question: It makes sense that we can't ship this before 12.0.0 but then what branch should I make the PR to?

`main`, but it won't be merged until around November

**Comment by anasik** [comment]
@AndriySvyryd got it. So for now, I should make a PR to `main` and ignore the #38746 and then when we merge my PR in November, we will address #38746?
